src/core/control/jobs/RenderJob.cpp

```diff
@@ -25,54 +26,34 @@ RenderJob::RenderJob(XojPageView* view): view(view) {}
 auto RenderJob::getSource() -> void* { return this->view; }
 
 void RenderJob::rerenderRectangle(Rectangle<double> const& rect) {
-    double zoom = view->xournal->getZoom();
-    Document* doc = view->xournal->getDocument();
+    const double ratio = view->xournal->getZoom() * this->view->xournal->getDpiScaleFactor();
 
     /**
      * Make sure the mask is big enough
-     * The +1 covers examples like rect.x = 0.4, rect.width = 1 and zoom = 1
+     * The +1 covers examples like rect.x = 0.4, rect.width = 1 and ratio = 1
      * We need a mask of width 2 pixels for that...
      **/
-    auto x = int(std::floor(rect.x * zoom));
-    auto y = int(std::floor(rect.y * zoom));
-    auto width = int(std::ceil(rect.width * zoom)) + 1;
-    auto height = int(std::ceil(rect.height * zoom)) + 1;
+    const auto x = int(std::floor(rect.x * ratio));
+    const auto y = int(std::floor(rect.y * ratio));
+    const auto width = int(std::ceil(rect.width * ratio)) + 1;
+    const auto height = int(std::ceil(rect.height * ratio)) + 1;
```

Comment:
1st. Too model, that we require at least a mask of two would be done in regular with a std::max(2, value).
2nd `cairo_surface_set_device_offset` only accepts doubles, why do we cast this here first into na int and then back to an double?

Reply:
It's not about size 2, this is just a dumb example to explain the +1. The same issue appears with anything like rect.x = n + 0.9 and rect.width = m + 0.2 (with m and n integers). In this situation, we need a width of m+2.
I'll make the comment more precise.

About the casting: you're right, it's not necessary for x and y. Only the fact that they are integer values matters, to avoid unnecessary antialiasing.